fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java

```diff
@@ -633,28 +633,38 @@ public synchronized List<Long> seqChooseBackendIds(int backendNum, boolean needA
             lastBackendId = lastBackendIdForOther;
         }
 
-        // host -> BE list
-        Map<String, List<Backend>> backendMaps = Maps.newHashMap();
-        for (Backend backend : srcBackends) {
-            // If needAvailable is true, unavailable backend won't go into the pick list
-            if (needAvailable && !backend.isAvailable()) {
-                continue;
+        List<Backend> backends = Lists.newArrayList();
+        if (Config.enable_replicas_on_the_same_host) {
+            for (Backend backend : srcBackends) {
+                // If needAvailable is true, unavailable backend won't go into the pick list
+                if (needAvailable && !backend.isAvailable()) {
+                    continue;
+                }
+                backends.add(backend);
             }
+        } else {
+            // host -> BE list
+            Map<String, List<Backend>> backendMaps = Maps.newHashMap();
+            for (Backend backend : srcBackends) {
+                // If needAvailable is true, unavailable backend won't go into the pick list
+                if (needAvailable && !backend.isAvailable()) {
+                    continue;
+                }
 
-            if (backendMaps.containsKey(backend.getHost())) {
-                backendMaps.get(backend.getHost()).add(backend);
-            } else {
-                List<Backend> list = Lists.newArrayList();
-                list.add(backend);
-                backendMaps.put(backend.getHost(), list);
+                if (backendMaps.containsKey(backend.getHost())) {
```

Comment:
```
                if (backendMaps.containsKey(backend.getHost())) {
                    backendMaps.get(backend.getHost()).add(backend);
                } else {
                    List<Backend> list = Lists.newArrayList();
                    list.add(backend);
                    backendMaps.put(backend.getHost(), list);
                }
```
use {put,compute}IfAbsent instead of a tedious block  of code.